### PR TITLE
Phase 2: Carmen Refactor

### PR DESCRIPTION
**Pull request by aarthykc**
### Context
Creating a clearer directory structure for carmen and carmen tests, since: 
- It makes code easier to navigate for new people (or, frankly, people who just haven’t seen it in a while)
- More obvious what you’re impacting with a change: indexing? geocoding? both?
- Ideally, makes it easier to identify other opportunities to DRY up code or do deeper refactors
- Troubleshooting: unwanted behaviours are easier to map to likely culprits

### Summary of Changes
- [ ] There will be many, will post big picture directory structure closer to when @boblannon and I ship this PR

### Next Steps
- [x] More code reorganisation
- [x] Get tests to pass
- [x] Revisit documentation to see if any links are broken as a result of moving stuff around
- [x] A very thorough review
- [ ] Mergetown

cc @mapbox/geocoding-gang

### CODE REVIEW

**Comment by boblannon** [comment]
reorganized lib folder:

```
lib/
├── constants.js
├── geocoder
│   ├── addresscluster.js
│   ├── addressitp.js
│   ├── context.js
│   ├── dedupe.js
│   ├── filter-sources.js
│   ├── geocode.js
│   ├── ops.js
│   ├── phrasematch.js
│   ├── spatialmatch.js
│   └── verifymatch.js
├── indexer
│   ├── cxxcache.js
│   ├── dawg.js
│   ├── index.js
│   ├── indexdocs.js
│   ├── merge-worker.js
│   └── merge.js
├── sources
│   ├── api-mem.js
│   └── loader.js
├── text-processing
│   ├── closest-lang.js
│   ├── equivalent-languages.json
│   ├── fallback-display.json
│   ├── fallback-indexer.json
│   ├── remove-diacritics.js
│   ├── termops.js
│   └── token.js
└── util
    ├── addfeature.js
    ├── analyze.js
    ├── bbox.js
    ├── feature.js
    ├── grid.js
    ├── permute.js
    ├── proximity.js
    └── uniq.js

5 directories, 34 files
```

**Comment by aarthykc** [comment]
yayyy green !

**Comment by aarthykc** [comment]
@ingalls, @samely this is ready for your review! This refactor of carmen is rather overwhelming as it touches almost every file, but it's aimed at reorganising the directory structure to be able to navigate easily. Below is a diagram of how we wanted our directory structure to look like:

```
.
├── lib
│   ├── geocoder
│   │   ├── forward
│   │   └── reverse
│   ├── indexer
│   ├── text-processing
│   ├── sources
│   └── util
├── bin
├── bench
└── test
    ├── acceptance
    └── unit
        ├── geocoder
        │   ├── forward
        │   └── reverse
        ├── indexer
        ├── text-processing
        ├── sources
        └── util
```
Earlier on @boblannon posted a screenshot of the directory structure after we moved files around:

```
lib/
├── constants.js
├── geocoder
│   ├── addresscluster.js
│   ├── addressitp.js
│   ├── context.js
│   ├── dedupe.js
│   ├── filter-sources.js
│   ├── geocode.js
│   ├── ops.js
│   ├── phrasematch.js
│   ├── spatialmatch.js
│   └── verifymatch.js
├── indexer
│   ├── cxxcache.js
│   ├── dawg.js
│   ├── index.js
│   ├── indexdocs.js
│   ├── merge-worker.js
│   └── merge.js
├── sources
│   ├── api-mem.js
│   └── loader.js
├── text-processing
│   ├── closest-lang.js
│   ├── equivalent-languages.json
│   ├── fallback-display.json
│   ├── fallback-indexer.json
│   ├── remove-diacritics.js
│   ├── termops.js
│   └── token.js
└── util
    ├── addfeature.js
    ├── analyze.js
    ├── bbox.js
    ├── feature.js
    ├── grid.js
    ├── permute.js
    ├── proximity.js
    └── uniq.js

```

Here are some more details, about specific files that moved and the location they moved to, incase you'd like to cross check:

<details>

bin/
    - carmen-merge (from /scripts/carmen-merge)
    - carmen-index (from /scripts/carmen-index)
    - carmen-analyze (from /scripts/carmen-analyze)
    - carmen (from /scripts/carmen)
    - tokenize.js (from /scripts/tokenize)
lib/
  constants.js (from lib/constants.js)
  text-processing/
    - termops.js (from lib/util/termops.js)
    - remove-diacritics.js (from lib/util/remove-diacritics)
    - token.js (from lib/util/token.js)
    - closest-lang.js (from lib/util/closest-lang)
    - fallback-display.json (from lib/util/fallback-display.json)
    - fallback-indexer.json (from lib/util/fallback-indexer.json)
    - equivalent.json → equivalent-languages.json (from lib/util/equivalent.json)
  indexer/
    - addfeature.js (from lib/util/addfeature.js)
    - index.js (from lib/)
    - merge.js (from lib/merge.js)
    - merge-worker.js (from lib/merge-worker.js)
    - dawg.js (move from lib/util/dawg.js)
    - cxxcache.js (move from lib/util/caxxcache.js)
  geocoder/
    - lib/geocode.js
    - spatialmatch (from lib/spatialmatch.js)
    - phrasematch (from lib/phrasematch.js)
    - verifymatch (from lib/verifymatch.js)
    - context.js (from lib/context.js)
    - ops.js (from util/ops.js)
    - dedupe.js (from util/dedupe.js)
    - filter (from util/filter.js) (maybe rename to filter-sources?)
    - addressitp.js (from lib/pure/addressitp.js)
    - addresscluster.js (from lib/pure) 
  sources/
    - api-mem.js (from lib/api-mem.js)
    - loader.js (from lib/loader.js)
  util/
    - analyze.js (from lib/analyze.js)
    - uniq.js (from lib/util/uniq.js)
    - proximity.js (from lib/util/proximity.js)
    - permute.js (from lib/util/permute.js)
    - grid.js (from lib/util/grid.js)
    - feature.js (from lib/util/feature.js)
    - bbox.js (from lib/util/bbox.js)
test/
  bin.test.js (tests command-line scripts)
  acceptance/
    - geocode-unit.*
    - tests that treat carmen as a black box
    - query goes in, response must match fixture
  unit/
    - should mirror the structure of lib
    - testing individual functions explicitly
  geocoder/
      - spatialmatch.*
      - verifymatch.test.js
      - address.*
      - filter
      - dedupe
      - context*
  indexer/
      - index* (move from test/index*) 
      - dawg 
      - cache
      - addFeature
  sources/
  text-processing/
      - termops.*
      - token
      - replace (rename token-replace)
      - fallback 
      - diacritics
      - closest-lang
  util/
      - uniq
      - grid 
      - proximity.*
      - permute
      - feature.*
      - bbox.*
</details>

File: index.js [hunk 1]
```diff
@@ -506,14 +505,6 @@ Geocoder.prototype.analyze = function(source, callback) {
...
 };
 
-// Copy a source's index to another.
-Geocoder.prototype.copy = function(from, to, callback) {
-    this._open((err) => {
-        if (err) return callback(err);
-        copy(from, to, callback);
-    });
-};
-
```
Comment (boblannon) [inline index.js]: this change is more than just a reorg: we removed the copy function, its supporting libs, and the script that uses it.

**Comment by apendleton** [comment]
Sorry for not noticing earlier, and no need to block merge, but curious why `dawg.js` and `cxxcache.js` ended up in `indexer`? These both define datastructures that are constructed at index time and queried at geocoding time.

**Comment by boblannon** [comment]
most of the decisions were made based on the dependency graph. the stuff in "indexer" is all part of the subgraph with `lib/indexer/index.js` at its root. 

after re-organizing we did sortof come to question the names "geocoder" and "indexer" as good names for that level of structure.  the things in "indexer" are all related because they directly interface with the "index" broadly defined (meaning dawg and cache).  the things in "geocoder" are all specifically for interpreting queries, ranking results and formatting the output.

so while i think the split is valid, i agree that another name might be a good idea.  it's tough, because the word "index" is so overloaded in this library...

File: bench/contextVector.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 var fs = require('fs');
 var Benchmark = require('benchmark');
 var suite = new Benchmark.Suite();
-var context = require('../lib/context');
+var context = require('../lib/geocoder/context');
```
Comment (ingalls) [inline bench/contextVector.js]: We should add ESLint to the `bench` folder for consistency

Ref: https://github.com/mapbox/carmen/blob/master/package.json#L63

Comment (boblannon) [reply bench/contextVector.js]: yeah, that's true. i'll open a followup PR to do that, in case it creates any difficulties.

**Pull request merged**